cmd/kubeadm/app/util/config/cluster.go

```diff
@@ -44,6 +45,16 @@ func FetchInitConfigurationFromCluster(client clientset.Interface, w io.Writer,
 	fmt.Fprintf(w, "[%s] Reading configuration from the cluster...\n", logPrefix)
 	fmt.Fprintf(w, "[%s] FYI: You can look at this config file with 'kubectl -n %s get cm %s -oyaml'\n", logPrefix, metav1.NamespaceSystem, constants.KubeadmConfigConfigMap)
 
+	// Check the health of the API server first. This /healthz call should timeout in 32 seconds.
+	// If client.Discovery().RESTClient() is nil, the fake client is used, and that means we are dry-running.
+	if client.Discovery().RESTClient() != nil {
+		healthStatus := 0
+		client.Discovery().RESTClient().Get().AbsPath("/healthz").Do().StatusCode(&healthStatus)
+		if healthStatus != http.StatusOK {
+			return nil, errors.Errorf("the API Server is unhealthy; /healthz didn't return %q", "ok")
+		}
+	}
+
 	// Fetch the actual config from cluster
 	cfg, err := getInitConfigurationFromCluster(constants.KubernetesDir, client, newControlPlane)
```

Comment:
It might happen that if we have a load balancer in front, the `/healthz` check will go against a working backend, and the real `getInitConfigurationFromCluster` will end up in a failing backend (if the LB didn't yet realize it's down and it's still on its active pool).

I don't think we should have this ad-hoc logic that is also weak because we don't really know the topology. Isn't there a way to shield the real http requests instead of a test & get approach?

Reply:
that is a good point.

> I don't think we should have this ad-hoc logic that is also weak because we don't really know the topology. Isn't there a way to shield the real http requests (by retrying the real get, with reasonable timeouts) instead of a test & get approach?

i agree, i don't really like the approach in this PR.
i think what has to be done here is to construct a custom Get call using the REST client that has the proper, reasonably short timeout.

Reply:
added WIP. will give this is short when i get some time.

Reply:
@ereslibre i've updated the PR with usage of List().